adapters/cloudwatch/uploader.go

```diff
@@ -0,0 +1,210 @@
+package cloudwatch
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+)
+
+// Uploader receieves CloudwatchBatches on its input channel,
+// and sends them on to the AWS Cloudwatch Logs endpoint.
+type Uploader struct {
+	Input    chan Batch
+	svc      *cloudwatchlogs.CloudWatchLogs
+	tokens   map[string]string
+	debugSet bool
+}
+
+// NewUploader creates and returns a new Uploader for the current EC2 Region
+func NewUploader(adapter *Adapter) *Uploader {
+	region := adapter.Route.Address
+	if (region == "auto") || (region == "") {
+		if adapter.Ec2Region == "" {
+			log.Println("cloudwatch: ERROR - could not get region from EC2")
+		} else {
+			region = adapter.Ec2Region
+		}
+	}
+	debugSet := false
+	_, debugOption := adapter.Route.Options[`DEBUG`]
+	if debugOption || (os.Getenv(`DEBUG`) != "") {
+		debugSet = true
+		log.Println("cloudwatch: Creating AWS Cloudwatch client for region",
+			region)
+	}
+	awsLogLevel := aws.LogOff
+	if debugSet {
+		awsLogLevel = aws.LogDebugWithRequestRetries
+	}
+	uploader := Uploader{
+		Input:    make(chan Batch),
+		tokens:   map[string]string{},
+		debugSet: debugSet,
+		svc: cloudwatchlogs.New(session.New(),
+			&aws.Config{
+				Region:     aws.String(region),
+				MaxRetries: &adapter.maxRetries,
+				LogLevel:   &awsLogLevel,
+			}),
+	}
+	go uploader.Start()
+	return &uploader
+}
+
+// Start begins the ain loop for the Uploader- POSTs each batch to AWS Cloudwatch
+// Logs, while keeping track of the unique sequence token for each log stream.
+func (u *Uploader) Start() {
+	for batch := range u.Input {
+		if len(batch.Msgs) > 0 {
```

Comment:
OK thanks for the check. Perhaps well add an `else` later to try and figure out where the empty Batches come from.